### DISCUSSION TITLE
OAuth2 and OpenID Connect

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#1847] [General]
Currently Oqtane supports a traditional AuthN/AuthZ approach which relies upon standard .NET Core Identity and cookies. This approach provides a reliable and secure method for managing security in a single independent web applicaton. However a variety of users have expressed a desire for Oqtane to support a wide variety of additional AuthN/AuthZ capabilities such as alternate authentication providers, single-sign-on (SSO), and delegated authorization for additional client applications - all of which are valuable in modern enterprise environments. As a result, the next major version of Oqtane will be enhanced with new security capabilities.

The primary enhancement will be the addition of an authorization server to the framework. An authorization server provides support for token-based authentication and authorization. It can offer multiple authentication methods simultaneously, support SSO, and facilitate authorization for all types of applications - web, mobile, desktop, etc... The authorization server will be based on OAuth2 and OpenID Connect, both of which are industry standards.

The solution will need to conform to Oqtane's philosophy and support the various capabilities which already exist in the framework. Specifically:

- it must support multi-tenancy
- it must support Blazor Server and Blazor WebAssembly hosting models
- it must require zero configuration on a default install

OAuth2 and OpenID Connect are very complex protocols. As a result, it does not make sense to attempt to build a custom authorization server. Therefore, this is a scenario where an exception is justified in regards to the Oqtane philosophy related to minimizing third party dependencies. After reviewing the available options and in particular their licensing restrictions and compatibility with Oqtane, it appears the clear choice is to leverage OpenIddict.

OpenIddict (https://github.com/openiddict/openiddict-core) is a native .NET solution offered under a prmissive Apache 2.0 license. It provides a versatile solution for implementing an OpenID Connect server and token validation in .NET Core. It fully supports the code/implicit/hybrid flows, the client credentials/resource owner password grants and the device authorization flow. It also has industry adoption from established solutions such as Orchard Core.

At this point it may be useful to shed a bit more light on the role of an authorization server. It simply allows you to login with a standard username and password (or with some other external provider). Then, based on the OAuth2 and OpenID Connect flows which have been configured, a client application can basically ask "can I perform a task on behalf of this user?". To answer this question, the user is redirected to the authorization server. Once the user is authenticated and it is determined the client application is allowed to make requests on behalf of the user, an access token will be sent to the client application to be used in subsequent requests.

Many folks have been patiently waiting for these capabilities to be added to the framework. I would be interested to hear any questions, thoughts, or feedback.

**Comment by vzdesic** [comment]
**This is great** that this will be part of the next major version!

I personally think that OpenIddict is good choice, still I would like to hear what thoughts guided you to choose this option instead of **Azure AD B2C** which was part of the few previous discussions?

Is it also intention to have this implemented in a **plug-and-play** manner from the beginning, so that we can replace it with our custom implementation?

**Reply by sbwalker** [reply]
@vzdesic you may be interested in reading this blog ( https://kevinchalet.com/2021/05/24/asp-net-core-6-and-authentication-servers-the-real-bait-and-switch-is-not-the-one-you-think/ ) - it offers a variety of useful information about IdentityServer, Duande, Azure AD B2C, and OpenIddict.

In regards to plug-and-play I was imagining a solution where different authentication providers could be supported ( ie. social logins, Azure AD, etc... ) - however I was not imagining that the authorization server itself could be replaced with a different authorization server. Authorization servers are complex and vary greatly in their implementations, so I am not sure if it would be feasible to try and make it pluggable.

**Comment by curlyfro** [comment]
this is fantastic news!  I assume OpenIddict will be the default auth provider shipped with Oqtane?  also, what's the plan to map user roles and permissions from the auth to Oqtane?

**Reply by sbwalker** [reply]
Oqtane will retain its own internal user, role, and permission concepts. External auth systems will need to map their users and/or roles to Oqtane's internal concepts.

**Comment by hishamco** [comment]
@sbwalker what's the plan for supporting OpenIddict into Oqtane? May I help on this if it's possible. FYI OpenIddict has a success story in OrchardCore, I'm glad to see it again here in Oqtane

**Comment by vzdesic** [comment]
Do you plan to create demo OpenIddict server and expose it in a separate repo here?
